File: leetcode30daychallenge/MoveZeroes.java
import java.util.*;
import java.io.*;

/*
 * https://leetcode.com/explore/featured/card/30-day-leetcoding-challenge/528/week-1/3286/
 * Given an array nums, write a function to move all 0's to the end of it while maintaining the relative order of the non-zero elements.
 *
 * Maintain a pointer j to insert non zero entries into array.
 * Iterate from left - O(n)
 */
class MoveZeroes {

    public static int[] moveZeroes(int[] nums) {
        int j = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0) {

            } else {
                nums[j] = nums[i];
                j++;
            }
        }
        for (int i = j; i < nums.length; i++) {
            nums[i] = 0;
        }
        return nums;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String line = sc.nextLine();
        String[] tokens = line.split(" ");
        int[] array = new int[tokens.length];
        for (int i = 0; i < tokens.length; i++) {
            array[i] = Integer.parseInt(tokens[i]);
        }
        System.out.println(Arrays.toString(moveZeroes(array)));
        sc.close();
    }
}

File: leetcodeAprilChallenge/python/MoveZeroes.py
class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        i = 0
        j = 0
        while i < len(nums):
            if nums[i] != 0:
                nums[j] = nums[i]
                j += 1
            i += 1

        i = j
        while (i < len(nums)):
            nums[i] = 0
            i += 1

    def moveZeroes2(self, nums: List[int]) -> None:
        i = 0
        for n in nums:
            if n != 0:
                nums[i] = n
            i += 1
        while i < len(nums):
            nums[i] = 0
            i += 1
